posts/fantastic-simprocs.md

```diff
@@ -0,0 +1,198 @@
+A rewriting rule is characterised by its *left hand side* and *right hand side*.
+Eg for a lemma of the form `LHS = RHS`, this is `LHS` and `RHS` respectively.
+
+The second one is the **`simp` tactic**, which we will refer to simply as `simp`.
+When run on a goal `⊢ e`, `simp` iteratively looks for a subexpression of `e` that matches the left hand side of some rewriting rule, and replaces that subexpression with the right hand side of that rule.
+
+For example, here's the proof steps `simp` follows to close the goal `37 * (Nat.fib 0 + 0) = 0`
+```lean
+⊢ 37 * (Nat.fib 0 + 0) = 0
+⊢ 37 * (0 + 0) = 0
+⊢ 37 * 0 = 0
+⊢ 0 = 0
+⊢ True
+```
+
+> The order in which `simp` traverses an expression is relatively intricate.
+> As a simple approximation, simplification is performed from the inside-out.
+> See the next blog post for full details on the algorithm.
+
+> If you write `set_option trace.Meta.Tactic.simp true in example : MyGoal := by simp`, you will see the list of simplification steps `simp` performs on `Goal`.
+
+# What is a simproc?
+
+Earlier, we said "almost all rewriting rules are lemmas".
+What about rewriting rules that are not lemmas?
+This is where simprocs enter the picture.
+
+A simproc is a program which, given an expression `LHS`, computes a new expression `RHS` and finds a proof of `LHS = RHS` on the fly.
+
+The concept of a simproc is genuinely more powerful than that of a simp lemma.
+Even though the right hand side of a lemma changes depending on its left hand side, it only does so via *syntactic substitution*.
+
+For example, the lemma `Nat.mul_add (a b c : Nat) : a * (b + c) = a * b + a * c` can be specialized to have right hand side `0 * b + 0 * c`, `a * b + a * (c + d)`, etc...
+But all these have *shape* `_ * _ + _ * _` and eg no value of `a`, `b`, `c` will make it have shape `_ * _ + _ * _ + _ * _`.
+
+With this in mind, simprocs are *parametric simp lemmas*.
+
+## Everything is a simproc
+
+As a very neat unification of concepts, it turns out that `simp` only ever handles simprocs.
+Every simp lemma is turned into a simproc before being fed to `simp`.
+
+# Examples of simprocs
+
+In this section, we exemplify four simprocs through the following use cases:
+* Computation
+* Avoiding combinatorial explosion of lemmas
+* Performance optimisation
+
+## Computation:
+
+### The `Finset.Icc_ofNat_ofNat` simproc
+
+The `Finset.Icc_ofNat_ofNat` simproc is designed to take expressions of the form `Finset.Icc a b` where `a` and `b` are numerals, and simplify them to an explicit set.
+
+```lean
+example : Finset.Icc 1 0 = ∅ := by
+  simp only [Icc_ofNat_ofNat]
+
+example : Finset.Icc 1 1 = {1} := by
+  simp only [Icc_ofNat_ofNat]
+
+example : Finset.Icc 1 4 = {1, 2, 3, 4} := by
+  simp only [Icc_ofNat_ofNat]
+
+example : Finset.Icc a (a + 2) = {a, a + 1, a + 2} := by
+  -- fails: the bounds of the interval aren't numerals!
+  simp only [Icc_ofNat_ofNat]
+```
```

Comment:
imo this one is less compelling than the `Nat.reduceDvd` one, and it currently doesn't have the informal description that all the others do. I don't mind keeping it, but I think it should go after the reduceDvd one, and have some explanation of the metaprogram (which can be one/two lines)